Implement security group egress and ingress method

watermelo opened the pull request:
## Implement security group egress and ingress method

### Description
1, Implement security group egress and ingress method
2, Provide unit test cases.

### Status

- done, ready for review

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [ ] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

Commit:
Merge branch 'upstream' into trunk

Commit:
Merge branch 'upstream' into trunk

Commit:
Implement some security group methods

1, Implement security group egress and ingress methods
2, Provide unit test cases

Commit:
modify unit test

Kami commented on libcloud/test/compute/test_ecs.py:
```diff
@@ -940,6 +955,20 @@ def _create_public_ip_AllocatePublicIpAddress(self, method, url, body, headers):
+        return (httplib.OK, resp_body, {}, httplib.responses[httplib.OK])
+
+
+if __name__ == '__main__':
```
Looks like this some how accidentally ended up here and now it's duplicated twice :)

Should remove one set.

Kami commented on libcloud/compute/drivers/ecs.py:
```diff
@@ -1289,6 +1289,102 @@ def create_public_ip(self, instance_id):
             return findtext(resp.object, 'IpAddress',
                             namespace=self.namespace)
 
+    def ex_security_group_ingress_permission(self, group_id,
```
If possible, it would be great if you could make this method look more like the methods in the EC2 and ClouStack driver (same method name and signature if possible) - https://github.com/apache/libcloud/blob/trunk/libcloud/compute/drivers/ec2.py#L4052

Kami commented on libcloud/compute/drivers/ecs.py:
```diff
@@ -1289,6 +1289,102 @@ def create_public_ip(self, instance_id):
                             namespace=self.namespace)
 
+    def ex_security_group_ingress_permission(self, group_id,
+                                             ip_protocol, portrange,
```
Minor consistency thing, please use `port_range` or perhaps `from_port` and `to_port` if you are going to make it look more like the methods in EC2 and CloudStack drivers.

Kami commented on libcloud/compute/drivers/ecs.py:
```diff
@@ -1289,6 +1289,102 @@ def create_public_ip(self, instance_id):
+        if portrange is None:
+            raise AttributeError('portrange is required')
+        if source_cidrip is None:
+            source_cidrip = '0.0.0.0/0'
```
Instead of doing that, please just set the default value of the argument in the method signature - `def (...., source_cidr_ip='0.0.0.0/0')`

Kami commented:
Thanks - I've added some comments.

Kami commented:
When you get a chance, please also file an ICLA - https://libcloud.readthedocs.io/en/latest/development.html#contributing-bigger-changes

Commit:
Modified method name

1, Modified the method name the same as others drivers.
2, Remove the extra code

watermelo commented:
Thanks, I've modified this fualt.😊

watermelo commented:
👌, I would like to sign ILCA.

tonybaloney commented:
@watermelo if you follow the steps in https://libcloud.readthedocs.io/en/latest/development.html#contributing-bigger-changes you can complete the ICLA online.

Kami commented:
First thank you for your contribution.

This PR is very old and out of date now so I will go ahead and close it.

If it's still relevant, please sync your changes with the latest master / trunk and open a new PR. Thanks!
